main/src/library/Array.flix

```diff
@@ -57,10 +57,1390 @@ namespace Array {
+    /// Helper for `sameElements`.
+    ///
+    def sameElementsHelper(a: Array[a], b: Array[b], len: Int32, i: Int32): Bool & Impure =
+        if (i >= len)
+            true
+        else {
+            if (a[i] != b[i]) false else sameElementsHelper(a, b, len, i+1)
+        }
+
+    ///
+    /// Apply the effectful function `f` to all the elements in the array `a`.
+    ///
+    pub def foreach(f: a ~> Unit, a: Array[a]): Unit & Impure =
+        let len = length(a);
+        if (len < 1) () else foreachHelper(f, a, len, 0)
+
+    ///
+    /// Helper for `foreach`.
+    ///
+    def foreachHelper(f: a ~> Unit, a: Array[a], len: Int32, i: Int32): Unit & Impure =
+        if (i >= len)
+            ()
+        else {
+            f(a[i]);
+            foreachHelper(f, a, len, i + 1)
+        }
+
+    ///
+    /// Apply the effectful function `f` to all the elements in the array `a`.
+    ///
+    pub def foreachWithIndex(f: (a, Int32) ~> Unit, a: Array[a]): Unit & Impure =
+        let len = length(a);
+        if (len < 1) () else foreachWithIndexHelper(f, a, len, 0)
+
+    ///
+    /// Helper for `foreachWithIndex`.
+    ///
+    def foreachWithIndexHelper(f: (a, Int32) ~> Unit, a: Array[a], len: Int32, i: Int32): Unit & Impure =
+        if (i >= len)
+            ()
+        else {
+            f(a[i], i);
+            foreachWithIndexHelper(f, a, len, i + 1)
+        }
+
+    ///
+    /// Apply the effectful function `f` to all the elements in the array `a` and the accumulator `s`.
+    /// The function returns the final value of the accumulator `s`.
+    ///
+    pub def foreachAccum(f: (a, st) ~> st, s: st, a: Array[a]): st & Impure =
+        foreachAccumHelper(f, a, length(a), s, 0)
+
+    ///
+    /// Helper for `foreachAccum`.
+    ///
+    def foreachAccumHelper(f: (a, st) ~> st, a: Array[a], len: Int32, s: st, i: Int32): st & Impure =
+        if (i >= len)
+            s
+        else {
+            let s1 = f(a[i], s);
+            foreachAccumHelper(f, a, len, s1, i + 1)
+        }
+
+    ///
+    /// Returns a copy of `a` with the elements starting at index `i` replaced by `sub`.
+    ///
+    pub def updateSequence(i: Int32, sub: Array[a], a: Array[a]): Array[a] & Impure =
+        let end = i + length(sub);
+        let f = ix -> if (ix >= i && ix < end) sub[ix - i] else a[ix];
+        init(f, length(a))
+
+    ///
+    /// Update `a` with the elements starting at index `i` replaced by `sub`.
+    ///
+    pub def updateSequenceInPlace(i: Int32, sub: Array[a], a: Array[a]): Unit & Impure =
+        let end = i + length(sub);
+        let f = { (_,ix) ->
+            if (ix >= i && ix < end)
+                a[ix] = sub[ix-i]
+            else ()
+        };
+        foreachWithIndex(f, a)
+
+    ///
+    /// Render the array `a` as a String. Elements are rendered with the
+    /// function `f` and seperated with the string `sep`.
+    ///
+    pub def formatWith(f: a -> String, sep: String, a: Array[a]): String & Impure =
```

Comment:
Looks good to me. The signature also seems sufficiently general that we could add it to all other collection types, e.g. set and map.